src/Orleans.Runtime/Transactions/TransactionAgent.cs

```diff
@@ -1,13 +1,99 @@
 using Orleans.Runtime;
 using Orleans.Concurrency;
+using Orleans.Runtime.Configuration;
+using DateTime = System.DateTime;
 
 namespace Orleans.Transactions
 {
+    internal class TransactionAgentMetrics
+    {
+        //TPS in current monitor window
+        private const string BatchStartTransanctionsDeltaTPS = "TransactionAgent.BatchStartTransactions.Delta.TPS";
+        //avg latency in current monitor window
+        private const string AvgBatchStartTransactionsDeltaLatency = "TransactionAgent.BatchStartTransactions.Delta.AvgLatency";
+        private const string BatchCommitTransactionsDeltaTPS = "TransactionAgent.BatchCommitTransactions.Delta.TPS";
+        private const string AvgBatchCommitTransactionsDeltaLatency = "TransactionAgent.BatchCommitTransactions.Delta.AvgLatency";
+
+        internal int BatchStartTransactionsRequestCounter { get; set; }
+
+        internal int BatchCommitTransactionsRequestsCounter { get; set; }
+        private TimeSpan batchStartTransactionsRequestLatencyCounter = TimeSpan.Zero;
+        internal TimeSpan BatchStartTransactionsRequestLatencyCounter
+        {
+            get { return batchStartTransactionsRequestLatencyCounter; }
+            set
+            {
+                batchStartTransactionsRequestLatencyCounter = value;
+                this.periodicMonitor.TryAction(DateTime.UtcNow);
+            }
+        }
+        private TimeSpan batchCommitTransactionsRequestLatencyCounter = TimeSpan.Zero;
+
+        internal TimeSpan BatchCommitTransactionsRequestLatencyCounter
+        {
+            get { return batchCommitTransactionsRequestLatencyCounter; }
+            set
+            {
+                batchCommitTransactionsRequestLatencyCounter = value;
+                this.periodicMonitor.TryAction(DateTime.UtcNow);
+            }
+        }
+        private DateTime lastReportTime = DateTime.UtcNow;
+        private ITelemetryProducer telemetryProducer;
+        private PeriodicAction periodicMonitor;
+
+        public TransactionAgentMetrics(ITelemetryProducer producer, TimeSpan interval)
+        {
+            this.telemetryProducer = producer;
+            this.periodicMonitor = new PeriodicAction(interval, this.ReportMetrics);
+        }
+
+        private void ResetCounters(DateTime lastReportTimeStamp)
+        {
+            //record last report time stamp
+            lastReportTime = lastReportTimeStamp;
+            this.BatchStartTransactionsRequestCounter = 0;
+            this.BatchCommitTransactionsRequestsCounter = 0;
+            this.batchStartTransactionsRequestLatencyCounter = TimeSpan.Zero;
+            this.batchCommitTransactionsRequestLatencyCounter = TimeSpan.Zero;
+        }
+
+        private void ReportMetrics()
+        {
+            var now = DateTime.UtcNow;
+           
+            //batch start metrics
+            var batchStartTransactionTPS = BatchStartTransactionsRequestCounter / (now - lastReportTime).TotalSeconds;
```

Comment:
Plz take max of total seconds and 1 to ensure no div by zero error

Reply:
why max? I can see have a zero check here , and skip if it is zero secs.
But take a max between total secs and 1 make the math wrong, isn't it?

Reply:
On the odd chance someone configures this wrong and it fires in less than a second, it will either report all the transactions in the window as TPS (which is incorrect for a window less than a second) or it will report nothing.
Either way works but yes, what I suggested will give wrong math, but only a little (in theory), whereas the alternative would produce no info a all.
Maybe we should ensure that the report interval is >= 1second?